server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -153,21 +239,39 @@ void doRun() {
                 continue;
             }
 
+            Class<? extends SearchBackpressureTask> taskType = getTaskType(taskCancellation.getTask());
+
             // Independently remove tokens from both token buckets.
-            boolean rateLimitReached = taskCancellationRateLimiter.get().request() == false;
-            boolean ratioLimitReached = taskCancellationRatioLimiter.get().request() == false;
+            boolean rateLimitReached = rateLimiters.get(taskType).get().request() == false;
+            boolean ratioLimitReached = ratioLimiters.get(taskType).get().request() == false;
 
             // Stop cancelling tasks if there are no tokens in either of the two token buckets.
             if (rateLimitReached && ratioLimitReached) {
                 logger.debug("task cancellation limit reached");
-                state.incrementLimitReachedCount();
+                SearchBackpressureState searchBackpressureState = searchBackpressureStates.get(taskType);
+                if (searchBackpressureState != null) {
+                    searchBackpressureState.incrementLimitReachedCount();
+                }
                 break;
             }
 
             taskCancellation.cancel();
```

Comment:
While cancelling co-ordinator task, I believe child tasks automatically get cancelled since `shouldCancelChildrenOnCancellation()` is true for `SearchTask`. Please verify this once with testing. Kindly ignore if already verified.

Reply:
Great point!

Using `TaskManager#cancelTaskAndDescendants` will allow us to cancel a parent task along with all its descendants, which will be useful for SearchTask cancellation at the coordinator level.

This wasn't required for SearchShardTask cancellation before, but is essential now. https://github.com/opensearch-project/OpenSearch/issues/4766

Reply:
It makes sense to use `TaskManager#cancelTaskAndDescendants` for parent task cancellation. Modified code to accommodate the same.